CanonicalABI.md

```diff
@@ -0,0 +1,1166 @@
+  be a WebAssembly instance or the host itself.
+* `ty` - the interface type that we're lifting into
+* `values` - the core wasm values iterator we have from the function signature
+  or results that we're lifting from. Note that due to type-checking and the
+  canonical ABI for function signatures it's guaranteed that `values` has the
+  right types to lift into `ty`.
+
+This defines how an interface types value is decoded from linear memory and
+WebAssembly values. Note that this function is modeled as a generator. This is
+used later during the definition of `fuse` to model how lifting/lowering is
+interleaved.
+
+```
+def lift(direction, src, ty, values):
+  match ty
+    f32 => yield(values.assert_next(wasm_f32)),
+    f64 => yield(values.assert_next(wasm_f64)),
+
+    s8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -128 && val <= 127)
+      yield(val as s8)
+    }
+
+    s16 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= -32768 && val <= 32767)
+      yield(val as s16)
+    }
+
+    u8 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= 0 && val <= 255)
+      yield(val as u8)
+    }
+    u16 => {
+      val = values.assert_next(wasm_i32)
+      assert_or_trap(val >= 0 && val <= 65535)
+      yield(val as u16)
+    }
+
+    s32 => yield(values.assert_next(wasm_i32) as s32),
+    u32 => yield(values.assert_next(wasm_i32) as u32),
+    s64 => yield(values.assert_next(wasm_i64) as s64),
+    u64 => yield(values.assert_next(wasm_i64) as u64),
+
+    char => {
+      codept = values.assert_next(wasm_i32)
+      # generate a trap if `codept` is out of bounds
+      assert_or_trap(codept < 0xD800 || (codpt >= 0xE000 && codept <= 0x10FFFF))
+      yield(codept as char)
+    }
+
+    # Note that lists are only deallocated during lowering for exports. For
+    # imports it's assumed the memory is still owned by the caller WebAssembly
+    # module.
+    #
+    # Also note that `read_utf8_string` here is expected to validate that the
+    # input read is indeed valid utf-8. One option is to trap on invalid utf-8,
+    # but another option is to match `TextDecoder.decode` on the web with
+    # replacement error mode which produces replacement characters.
+    #
+    # Currently it's expected taht `read_utf8_string` will not trap and will
+    # instead follow the `TextDecoder.decode` semantics which is what WTF-16
+    # languages like JS/Java/C# will want.
+    string => {
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+      assert_or_trap(ptr + len <= src.memory.len)
+
+      val = read_utf8_string(src, ptr, len)
+      if direction == "export":
+        src.canonical_abi_free(ptr, len, 1)
+      yield(val as string)
+    }
+
+    list<$ty> => {
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      assert_or_trap(ptr % align(direction, $ty) == 0)
```

Comment:
I'm now wondering whether the canonical ABI should perhaps not `assert_or_trap` on alignment of linear memory allocations.  My reasoning is:
* `memory.copy` et al don't
* if the canonical ABI impl wants to call some optimized routine that requires alignment, the same branch needed for trapping could instead redirect to a byte-unaligned version.  (Although, practically speaking, I expect canonical ABI impls will just use `memcpy` to handle it for them.)

Thus, requiring alignment might just be making work for producer tools without providing much in the way of benefit for perf or engines.  Am I missing any other reasons?

Reply:
The main downside which made me think that alignment should be respected is that all language bindings will need to use unaligned things I think? For example it's unsound in Rust to have `&T` where the memory is actually misaligned (due to LLVM optimizations), so I figured since everyone will respect alignment anyway we may as well go ahead and guarantee it in the ABI.

If we can solve that issue though it seems reasonable to not require alignment I think. Now that I think about it, because the module itself is always responsible for managing all the pointer (either they come from the source or from malloc), perhaps my concern here is a moot point? Rust can guarantee that everything is still correctly aligned but from the canonical abi perspective it's not inventing pointers out of thin air and it'll work with unaligned things if you pass it in.

I do think, though, we'd still want to pass the alignment to realloc/free, do you think that's ok?

Reply:
Yes, totally agreed.